Resources/NetHook2/NetHook2/net.cpp

```diff
@@ -0,0 +1,120 @@
+
+#define WIN32_LEAN_AND_MEAN
+#include <windows.h>
+
+
+#include "net.h"
+
+#include "logger.h"
+#include "csimplescan.h"
+
+
+namespace NetHook
+{
+
+
+BBuildAndAsyncSendFrameFn BBuildAndAsyncSendFrame_Orig = NULL;
+RecvPktFn RecvPkt_Orig = NULL;
+
+CNet::CNet()
+	: m_RecvPktDetour(NULL),
+	  m_BuildDetour(NULL)
+{
+	CSimpleScan steamClientScan("steamclient.dll");
+
+	BBuildAndAsyncSendFrameFn pBuildFunc = NULL;
+	bool bFoundBuildFunc = steamClientScan.FindFunction(
+		"\x55\x8B\xEC\x83\xEC\x6C\x53\x8B\xD9\x89\x5D\xF4",
+		"xxxxxxxxxxxx",
+		(void **)&pBuildFunc
+	);
+
+	BBuildAndAsyncSendFrame_Orig = pBuildFunc;
+
+	g_pLogger->LogConsole("CWebSocketConnection::BBuildAndAsyncSendFrame = 0x%x\n", BBuildAndAsyncSendFrame_Orig);
+
+	RecvPktFn pRecvPktFunc = NULL;
+	bool bFoundRecvPktFunc = steamClientScan.FindFunction(
+		"\x55\x8B\xEC\x81\xEC\xA0\x09\x00\x00\x53",
+		"xxxxxxxxxx",
+		(void **)&pRecvPktFunc
+	);
+
+	RecvPkt_Orig = pRecvPktFunc;
+
+	g_pLogger->LogConsole("CCMInterface::RecvPkt = 0x%x\n", RecvPkt_Orig);
+
+
+	RecvPktFn thisRecvPktFunc = CNet::RecvPkt;
+
+	if (bFoundBuildFunc)
+	{
+		BBuildAndAsyncSendFrameFn thisBuildFunc = CNet::BBuildAndAsyncSendFrame;
+
+		m_BuildDetour = new CSimpleDetour((void **)&BBuildAndAsyncSendFrame_Orig, (void *)thisBuildFunc);
+		m_BuildDetour->Attach();
+
+		g_pLogger->LogConsole("Detoured CWebSocketConnection::BBuildAndAsyncSendFrame!\n");
+	}
+	else
+	{
+		g_pLogger->LogConsole("Unable to hook CWebSocketConnection::BBuildAndAsyncSendFrame: func scan failed.\n");
+	}
+
+	if (bFoundRecvPktFunc)
+	{
+		RecvPktFn thisRecvPktFunc = CNet::RecvPkt;
+
+		m_RecvPktDetour = new CSimpleDetour((void **)&RecvPkt_Orig, (void *)thisRecvPktFunc);
+		m_RecvPktDetour->Attach();
+
+		g_pLogger->LogConsole("Detoured CCMInterface::RecvPkt!\n");
+	}
+	else
+	{
+		g_pLogger->LogConsole("Unable to hook CCMInterface::RecvPkt: func scan failed.\n");
+	}
+
+}
+
+CNet::~CNet()
+{
+	if (m_RecvPktDetour)
+	{
+		m_RecvPktDetour->Detach();
+		delete m_RecvPktDetour;
+	}
+
+	if (m_BuildDetour)
+	{
+		m_BuildDetour->Detach();
+		delete m_BuildDetour;
+	}
+}
+
+
+bool CNet::BBuildAndAsyncSendFrame(void *webSocketConnection, void *unused, EWebSocketOpCode eWebSocketOpCode, const uint8 *pubData, uint32 cubData)

```

Comment:
Is `unused` actually unused, or just unknown?

Reply:
`fastcall` is just a quick and dirty way to hook `thiscall` methods on windows since it passes the first 2 parameters through `ecx` and `edx`, and we really only care about the `this` pointer being passed through `ecx`. `unused` is just the value in `edx`.

Reply:
Nasty, but clever.